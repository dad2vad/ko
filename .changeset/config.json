{
  "$schema": "https://unpkg.com/@changesets/config@1.5.0/schema.json",
  "changelog": ["@changesets/changelog-github", { "repo": "tommy351/kosko" }],
  "commit": false,
  "linked": [["@kosko/cli", "kosko"]],
  "access": "public",
  "baseBranch": "main",
  "updateInternalDependencies": "patch",
  "ignore": []
}
