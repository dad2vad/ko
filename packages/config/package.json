{
  "name": "@kosko/config",
  "version": "3.0.1",
  "description": "Load kosko configuration.",
  "homepage": "https://kosko.dev",
  "repository": "https://github.com/tommy351/kosko/tree/master/packages/config",
  "author": "Tommy Chen <tommy351@gmail.com>",
  "license": "MIT",
  "main": "dist/index.node.cjs",
  "module": "dist/index.node.mjs",
  "types": "dist/types.d.ts",
  "sideEffects": false,
  "scripts": {
    "build": "build-scripts"
  },
  "exports": {
    "types": "./dist/index.d.ts",
    "deno": "./dist/index.deno.mjs",
    "node": {
      "import": "./dist/index.node.mjs",
      "require": "./dist/index.node.cjs"
    },
    "import": "./dist/index.base.mjs",
    "require": "./dist/index.node.cjs"
  },
  "files": [
    "/dist/"
  ],
  "keywords": [
    "kosko"
  ],
  "publishConfig": {
    "access": "public"
  },
  "engines": {
    "node": ">=14.18.0"
  },
  "dependencies": {
    "@iarna/toml": "^2.2.5",
    "@kosko/common-utils": "workspace:^",
    "@kosko/log": "workspace:^",
    "superstruct": "^0.15.3"
  },
  "devDependencies": {
    "@kosko/build-scripts": "workspace:^",
    "@kosko/jest-preset": "workspace:^"
  },
  "jest": {
    "preset": "@kosko/jest-preset"
  }
}
