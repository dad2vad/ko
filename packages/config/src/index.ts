/**
 * Contains config types and loaders.
 *
 * @packageDocumentation
 */

export * from "./config";
export type { EnvironmentConfig, Config } from "./types";
export * from "./validate";
