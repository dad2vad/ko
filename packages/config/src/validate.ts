import { validate as sValidate, Failure } from "superstruct";
import { Config, configSchema } from "./types";

function formatErrors(failures: readonly Failure[]): string {
  return [
    'Config file "kosko.toml" is invalid.',
    ...failures.map((f) => `- "${f.path.join(".")}": ${f.message}`)
  ].join("\n");
}

/**
 * @public
 */
export class ValidationError extends Error {
  /**
   * An array of {@link https://docs.superstructjs.org/api-reference/errors | superstruct error objects}.
   */
  public readonly errors: readonly Failure[];

  constructor(errors: readonly Failure[]) {
    super(formatErrors(errors));

    this.errors = errors;
  }
}

ValidationError.prototype.name = "ValidationError";

/**
 * Validates data against the configuration schema. It throws a {@link ValidationError}
 * when validation failed.
 *
 * @public
 */
export function validate(data: unknown): Config {
  const result = sValidate(data, configSchema);

  if (result[0]) {
    throw new ValidationError(result[0].failures());
  }

  return result[1];
}
