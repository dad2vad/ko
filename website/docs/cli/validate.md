---
title: kosko validate
---

Validate components.

This command is almost identical to [`kosko generate`](generate.md) command, but does not print Kubernetes manifests to terminal. You can see more details and examples there.

```shell
kosko validate [components..]
```
