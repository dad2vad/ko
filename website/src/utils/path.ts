export const sep = "/";

export function isAbsolute(path: string): boolean {
  return path.startsWith(sep);
}

export function isRelative(path: string): boolean {
  return path.startsWith(`.${sep}`) || path.startsWith(`../${sep}`);
}

export function dirname(path: string): string {
  const index = path.lastIndexOf(sep);
  if (index === -1) return ".";
  return path.substring(0, index) || sep;
}

export function extname(path: string): string {
  const index = path.lastIndexOf(".");
  if (index === -1) return "";
  return path.substring(index);
}

export function basename(path: string, ext?: string): string {
  const index = path.lastIndexOf(sep);
  const base = index === -1 ? path : path.substring(index + 1);

  if (base.endsWith(ext)) {
    return base.substring(0, base.length - ext.length);
  }

  return base;
}

export function relative(from: string, to: string): string {
  const fromParts = from.split(sep).filter(Boolean);
  const toParts = to.split(sep).filter(Boolean);

  while (fromParts[0] && toParts[0]) {
    fromParts.shift();
    toParts.shift();
  }

  while (toParts[0] === "." || toParts[0] === "..") {
    const toPart = toParts.shift();

    if (toPart === "..") {
      fromParts.pop();
    }
  }

  while (fromParts.pop()) {
    toParts.unshift("..");
  }

  return toParts.join(sep);
}
