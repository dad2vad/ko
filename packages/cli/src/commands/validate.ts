import { Command } from "../cli/command";
import logger, { LogLevel } from "@kosko/log";
import { handler } from "./generate/worker";
import { BaseGenerateArguments } from "./generate/types";
import { generateBuilder } from "./generate/command";
import { loadConfig } from "./generate/config";

export type ValidateArguments = BaseGenerateArguments;

export const validateCmd: Command<ValidateArguments> = {
  command: "validate [components..]",
  describe: "Validate components",
  builder(argv) {
    /* istanbul ignore next */
    return generateBuilder(argv)
      .example("$0 validate", "Validate components")
      .example("$0 validate foo bar", "Specify components")
      .example("$0 validate foo_*", "Use glob syntax");
  },
  async handler(args) {
    const config = await loadConfig(args);

    await handler({
      args: { ...args, validate: true },
      config
    });

    logger.log(LogLevel.Info, "Components are valid");
  }
};
