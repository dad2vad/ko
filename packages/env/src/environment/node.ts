import { getRequireExtensions } from "@kosko/require";
import { join } from "node:path";
import { formatPath } from "../paths";
import { toArray } from "@kosko/common-utils";
import {
  createEnvironment,
  createLoaderReducers,
  createReducerList
} from "./base";
import { Environment, Loader, ReducerExecutor, ReducerList } from "./types";
import { cwd as getCwd } from "node:process";

export function getNodeExtensions(): string[] {
  return getRequireExtensions().map((ext) => ext.substring(1));
}

/**
 * Options for creating a Node.js environment.
 *
 * @public
 */
export interface NodeEnvironmentOptions {
  /**
   * Current working directory (CWD).
   *
   * @defaultValue `process.cwd()`
   */
  cwd?: string;
}

export interface InternalNodeEnvironmentOptions extends NodeEnvironmentOptions {
  createReducerExecutor(this: void, reducers: ReducerList): ReducerExecutor;
  requireModule(this: void, environment: Environment, id: string): any;
  mergeValues(this: void, data: any[]): any;
}

export function createNodeEnvironment({
  cwd = getCwd(),
  createReducerExecutor,
  requireModule,
  mergeValues
}: InternalNodeEnvironmentOptions): Environment {
  function requireAllEnvs(template: string, component?: string) {
    if (!environment.env) return [];

    const envs = toArray(environment.env);

    return envs.map((env) => {
      const path = formatPath(template, {
        environment: env,
        ...(component && { component })
      });

      return requireModule(environment, join(environment.cwd, path));
    });
  }

  const loader: Loader = {
    global: () => requireAllEnvs(environment.paths.global),
    component: (name) => requireAllEnvs(environment.paths.component, name)
  };
  const reducers = createReducerList(createLoaderReducers(loader, mergeValues));
  const { reduce } = createReducerExecutor(reducers);
  const environment = createEnvironment({
    cwd,
    extensions: getNodeExtensions(),
    setReducers: reducers.setReducers,
    resetReducers: reducers.resetReducers,
    reduce
  });

  return environment;
}
