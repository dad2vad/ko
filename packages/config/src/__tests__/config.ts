import { join } from "node:path";
import { getConfig, loadConfig, searchConfig } from "../config";
import { ValidationError } from "../validate";

const fixturePath = join(__dirname, "..", "__fixtures__");

describe("loadConfig", () => {
  describe("when file exists", () => {
    test("should load the config", async () => {
      const config = await loadConfig(join(fixturePath, "toml", "kosko.toml"));
      expect(config).toMatchSnapshot();
    });
  });

  describe("when file does not exist", () => {
    test("should throw an error", async () => {
      await expect(loadConfig(join(fixturePath, "nowhere"))).rejects.toThrow(
        "ENOENT"
      );
    });
  });

  describe("when config is invalid", () => {
    test("should throw ValidationError", async () => {
      await expect(
        loadConfig(join(fixturePath, "invalid", "kosko.toml"))
      ).rejects.toThrow(ValidationError);
    });
  });
});

describe("searchConfig", () => {
  test("should return the config when kosko.toml exists", async () => {
    const config = await searchConfig(join(fixturePath, "toml"));
    expect(config).toMatchSnapshot();
  });

  test("should return an empty object when kosko.toml does not exist", async () => {
    const config = await searchConfig(join(fixturePath, "nowhere"));
    expect(config).toEqual({});
  });

  test("should throw ValidationError when config is invalid", async () => {
    const cwd = join(fixturePath, "invalid");
    await expect(searchConfig(cwd)).rejects.toThrow(ValidationError);
  });
});

describe("getConfig", () => {
  test("should return an empty config when config is empty", () => {
    expect(getConfig({}, "dev")).toEqual({
      components: [],
      require: [],
      loaders: []
    });
  });

  test("should return the global config when env is not defined", () => {
    const input = {
      components: ["foo"],
      require: ["bar"],
      loaders: ["baz"]
    };

    expect(getConfig(input, "dev")).toEqual(input);
  });

  test("should return the global config when env is defined but the key does not exist", () => {
    const globalConf = {
      components: ["foo"],
      require: ["bar"],
      loaders: ["baz"]
    };
    const input = {
      ...globalConf,
      environments: {
        prod: {
          components: ["aaa"],
          require: ["bbb"],
          loaders: ["ccc"]
        }
      }
    };

    expect(getConfig(input, "dev")).toEqual(globalConf);
  });

  test("should return the merge of global and environment config when env is defined and the key exists", () => {
    const input = {
      components: ["foo"],
      require: ["bar"],
      loaders: ["baz"],
      environments: {
        dev: {
          components: ["aaa"],
          require: ["bbb"],
          loaders: ["ccc"]
        }
      }
    };

    expect(getConfig(input, "dev")).toEqual({
      components: ["foo", "aaa"],
      require: ["bar", "bbb"],
      loaders: ["baz", "ccc"]
    });
  });

  test("should return the merge of multiple envs when env is an array", () => {
    const input = {
      components: ["foo"],
      require: ["bar"],
      loaders: ["baz"],
      environments: {
        a: {
          components: ["aa"],
          require: ["ab"],
          loaders: ["ac"]
        },
        c: {
          components: ["ca"],
          require: ["cb"],
          loaders: ["cc"]
        }
      }
    };

    expect(getConfig(input, ["a", "b", "c"])).toEqual({
      components: ["foo", "aa", "ca"],
      require: ["bar", "ab", "cb"],
      loaders: ["baz", "ac", "cc"]
    });
  });
});
