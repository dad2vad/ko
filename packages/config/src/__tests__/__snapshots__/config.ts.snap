// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadConfig when file exists should load the config 1`] = `
{
  "bail": true,
  "baseEnvironment": "m",
  "components": [
    "b",
    "c",
  ],
  "environments": {
    "dev": {
      "components": [
        "f",
        "g",
      ],
      "loaders": [
        "p",
        "q",
      ],
      "require": [
        "d",
        "e",
      ],
    },
    "prod": {
      "components": [
        "i",
        "j",
      ],
      "loaders": [
        "r",
        "s",
      ],
      "require": [
        "h",
      ],
    },
  },
  "extensions": [
    "js",
    "json",
  ],
  "loaders": [
    "n",
    "o",
  ],
  "paths": {
    "environment": {
      "component": "l",
      "global": "k",
    },
  },
  "require": [
    "a",
  ],
}
`;

exports[`searchConfig should return the config when kosko.toml exists 1`] = `
{
  "bail": true,
  "baseEnvironment": "m",
  "components": [
    "b",
    "c",
  ],
  "environments": {
    "dev": {
      "components": [
        "f",
        "g",
      ],
      "loaders": [
        "p",
        "q",
      ],
      "require": [
        "d",
        "e",
      ],
    },
    "prod": {
      "components": [
        "i",
        "j",
      ],
      "loaders": [
        "r",
        "s",
      ],
      "require": [
        "h",
      ],
    },
  },
  "extensions": [
    "js",
    "json",
  ],
  "loaders": [
    "n",
    "o",
  ],
  "paths": {
    "environment": {
      "component": "l",
      "global": "k",
    },
  },
  "require": [
    "a",
  ],
}
`;
